### DISCUSSION TITLE
[Testing Automation] Enhancing 3DS card payment for Cybersource connector using Cypress

### DISCUSSION THREAD

**Discussion by neerajkumar-n** [juspay/hyperswitch#4315] [General]
**Context:**\
Ensuring the smooth operation of our payment processing system is crucial for the success of our application. One of the Connectors we rely on is Cybersource payment gateway and we use this for handling transactions, focusing on both one-time payments & Subscription/repeat transactions.

Our current testing framework lacks comprehensive coverage for both one-time & Subscription payment scenarios with the Cybersource connector. We currently use Newman for automation testing but have encountered challenges. Maintaining separate collections for different connectors has become burdensome, requiring manual verification of every flow for each connector with every change. Newman's inability to handle redirection flows, like 3DS flows, and its limitations in testing other payment methods such as wallets and bank transfers, further compound our testing efforts. Recognizing the inefficiency and limitations, we acknowledge the need to transition to Cypress.

By automating our testing processes using cypress, we aim to streamline workflows, improve efficiency, and address the shortcomings encountered with our current automation testing, ensuring a more robust and reliable testing strategy moving forward. We are committed to enhancing our testing strategy by incorporating dedicated test scenarios specifically targeting one-time payments within the Cybersource connector. By doing so, we aim to ensure that all aspects of one-time payments, including transaction initiation, authorization, and confirmation, are thoroughly tested and validated for card payments.

In summary, by prioritizing the enhancement of our testing framework to include one-time payment scenarios, we reaffirm our commitment to quality, reliability, and continuous improvement in our payment processing ecosystem. This approach will enable us to identify and rectify potential issues before they impact our users' experience.

****

**How to get started**

- [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md)

- Signup and creating a sandbox account on [Cybersource](https://developer.cybersource.com/hello-world/sandbox.html) to get the API Credentials

- Configure Connector within Hyperswitch dashboard, you can sign up for Hyperswitch [here](http://app.hyperswitch.io). You can learn more about configuring Connectors in our documentation [here](https://docs.hyperswitch.io/hyperswitch-cloud/account-setup).

- Once configured, you can simply test a payment via our dashboard, learn more about it [here](https://docs.hyperswitch.io/hyperswitch-open-source/testing/test-a-payment)

**How to Implement:**

- Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch

- Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3) 

- You can refer to our Documentation [here](http://api-reference.hyperswitch.io)

****

**Expected Outcomes:**

Testing Procedure for Card Payment Authentication (authentication\_type: three\_ds)
- Initiate Card Payment with Three-DS Authentication:
  - Create a card payment and confirm it with the authentication\_type set as "three\_ds".

- Verification of Response Link:
  - Verify the link provided in the response.
  - Open the provided link to proceed with authentication.

- Complete Authentication and Redirect:
  - Follow the authentication process and complete it.
  - Redirect back to the return URL specified.

- Verification of Payment Status:
  - Verify the status of the payment after authentication completion.

- Failed Authentication Handling:
  - Simulate a failed authentication attempt.
  - Verify that the payment status reflects the failed authentication state.

Submit a pull request to the juspay/hyperswitch repository including the relevant screenshots of the test cases.\
Brownie Points: Providing clear and comprehensive documentation and code comments.\
Note: We welcome multiple PRs for this problem statement

****

**Additional Resources:** 

- If you face any issues/challenges while using Cypress, we recommend you to refer to [their documentation](https://docs.cypress.io/guides/overview/why-cypress)

- Here is a [test case](https://github.com/juspay/hyperswitch/tree/main/cypress-tests) that we have already worked upon for your reference.

- You can read more about 3DS [here](https://www.ebanx.com/en/resources/payments-explained/3d-secure/)

**Comment by hp77-creator** [comment]
Just a question, why cypress? afaik Cypress's power comes up when we need to do E2E for web-applications, for API e2e wouldn't it have been better to go with something light?

**Reply by Gnanasundari24** [reply]
Hi , We chose Cypress because we want to use one tool for testing both web and API. While Cypress is best known for web app testing, we think it can handle API testing too, keeping things simple for us.

**Comment by neerajkumar-n** [comment]
Closing this discussion, this has been implemented.
